util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,411 @@
+		tsc.collect(input.timestamp, input.records)
+	}
+}
+
+// collect uses a hashmap to store records in every second, and evict when necessary.
+// This function can be run in parallel with snapshot, so we should protect the map operations with a mutex.
+func (tsc *TopSQLReporterImpl) collect(timestamp uint64, records []tracecpu.TopSQLCPUTimeRecord) {
+	for _, record := range records {
+		encodedKey := encodeCacheKey(record.SQLDigest, record.PlanDigest)
+		entry, exist := tsc.topSQLMap[string(encodedKey)]
+		if !exist {
+			entry = &topSQLDataPoints{
+				SQLDigest:     record.SQLDigest,
+				PlanDigest:    record.PlanDigest,
+				CPUTimeMsList: []uint32{record.CPUTimeMs},
+				TimestampList: []uint64{timestamp},
+			}
+			tsc.topSQLMap[string(encodedKey)] = entry
+		} else {
+			entry.CPUTimeMsList = append(entry.CPUTimeMsList, record.CPUTimeMs)
+			entry.TimestampList = append(entry.TimestampList, timestamp)
+		}
+		entry.CPUTimeMsTotal += uint64(record.CPUTimeMs)
+	}
+
+	if len(tsc.topSQLMap) <= tsc.MaxStatementsNum {
+		return
+	}
+
+	// find the max CPUTimeMsTotal that should be evicted
+	digestCPUTimeList := make([]cpuTimeSort, len(tsc.topSQLMap))
+	{
+		i := 0
+		for key, value := range tsc.topSQLMap {
+			data := cpuTimeSort{
+				Key:            key,
+				SQLDigest:      value.SQLDigest,
+				PlanDigest:     value.PlanDigest,
+				CPUTimeMsTotal: value.CPUTimeMsTotal,
+			}
+			digestCPUTimeList[i] = data
+			i++
+		}
+	}
+
+	// QuickSelect will only return error when the second parameter is out of range
+	if err := quickselect.QuickSelect(cpuTimeSortSlice(digestCPUTimeList), tsc.MaxStatementsNum); err != nil {
+		//	skip eviction
+		return
+
+	}
+
+	// TODO: we can change to periodical eviction (every minute) to relax the CPU pressure
+	shouldEvictList := digestCPUTimeList[tsc.MaxStatementsNum:]
+	for _, evict := range shouldEvictList {
+		delete(tsc.topSQLMap, evict.Key)
+		tsc.mu.Lock()
+		delete(tsc.normalizedSQLMap, string(evict.SQLDigest))
+		delete(tsc.normalizedPlanMap, string(evict.PlanDigest))
+		tsc.mu.Unlock()
+	}
+}
+
+// RegisterSQL registers a normalized SQL string to a SQL digest.
+//
+// Note that the normalized SQL string can be of >1M long.
+// This function should be thread-safe, which means parallelly calling it in several goroutines should be fine.
+// It should also return immediately, and do any CPU-intensive job asynchronously.
+// TODO: benchmark test concurrent performance
+func (tsc *TopSQLReporterImpl) RegisterSQL(sqlDigest string, normalizedSQL string) {
+	tsc.mu.RLock()
+	_, exist := tsc.normalizedSQLMap[sqlDigest]
+	tsc.mu.RUnlock()
+	if !exist {
+		tsc.mu.Lock()
+		tsc.normalizedSQLMap[sqlDigest] = normalizedSQL
+		tsc.mu.Unlock()
+	}
+}
+
+// RegisterPlan is like RegisterSQL, but for normalized plan strings.
+// TODO: benchmark test concurrent performance
+func (tsc *TopSQLReporterImpl) RegisterPlan(planDigest string, normalizedPlan string) {
+	tsc.planRegisterChan <- &planRegisterJob{
```

Comment:
If we don't want it to be blocked then the buffer size should be enlarged.

Reply:
Changed to 1024.